### PR TITLE
VMSS build agent + CIS compliance fixes

### PR DESCRIPTION
**Pull request by t-young31**
## Partially resolves https://github.com/UCLH-Foundry/FlowEHR/issues/199

### CODE REVIEW

File: deployment/build-agent/cloud-config.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  - azure-cli
+  - gcc
+
+runcmd:
+  - export DEBIAN_FRONTEND=noninteractive
+  - mkdir -p /tmp/actions-runner
+  - curl -o runner.tar.gz -L https://github.com/actions/runner/releases/download/${GITHUB_RUNNER_VERSION}/actions-runner-linux-x64-2.302.1.tar.gz
+  - tar xzf runner.tar.gz -C /tmp/actions-runner
+  - rm runner.tar.gz
+  - bash /start_gh_runner.sh
```
Comment (t-young31) [inline deployment/build-agent/cloud-config.yaml:82]: this should probably be run a service so it executes on boot

File: scripts/init.py
Comment (t-young31) [inline scripts/init.py:1]: @jjgriff93 what do you think of this? (think better to load it from a yaml?)

File: .gitignore [hunk 2]
```diff
@@ -23,3 +23,10 @@ override.tf.json
 
 # Include tfplan files to ignore the plan output of command: terraform plan -out=tfplan
 # example: *tfplan*
+
+# Exclude the top level config file
+config.tfvars
+
+# Exclude the terraform state in the public template repo
```
Comment (jjgriff93) [inline .gitignore:30]: Yeah was torn with this. Yeah perhaps we do gitignore it just so we don't accidentally check any state in but we need to have an instruction in the readme that says ideally remove the gitignore as you should check in state blah blah blah

File: deployment/build-agent/cloud-config.yaml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        | awk '/token/ { gsub(/[",]/,""); print $2}')
+
+      if [ "$REGISTRATION_TOKEN" != "" ]; then
+        echo "Created registration token"
+      else
+        echo "Failed to obtain a registration token. Check the scope of \$GITHUB_RUNNER_TOKEN"
+        exit 1
+      fi
+
+      export RUNNER_ALLOW_RUNASROOT="1"
```
[the remark this answers is not in the document]
Comment (jjgriff93) [reply deployment/build-agent/cloud-config.yaml:54]: Do we have a choice?

File: deployment/storage.tf [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+#  limitations under the License.
+
+resource "azurerm_storage_account" "bootstrap" {
+  name                              = "stgbtstr${local.suffix_truncated}"
+  resource_group_name               = azurerm_resource_group.bootstrap.name
+  location                          = azurerm_resource_group.bootstrap.location
+  account_tier                      = "Standard"
+  account_replication_type          = "GRS"
+  infrastructure_encryption_enabled = true
+  public_network_access_enabled     = true  # Turned off post apply
```
Comment (jjgriff93) [inline deployment/storage.tf:22]: Nice!

File: scripts/clean_terraform_state.sh [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#  limitations under the License.
+
+set -o errexit
+set -o pipefail
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+
+# Remove any admin passwords from the state
+STATE_FILEPATH="${SCRIPT_DIR}/../deployment/terraform.tfstate"
+content=$(jq 'del(.. | .admin_password?)' "$STATE_FILEPATH")
```
Comment (jjgriff93) [inline scripts/clean_terraform_state.sh:23]: Beautiful

File: .gitignore (hunk 2 above)
Comment (t-young31) [reply .gitignore:30]: why have an instruction when you could have [a script](https://github.com/UCLH-Foundry/Azure-Bootstrap/blob/t-young31/vmss_and_cis/scripts/init.py).. mostly for fools like me who don't read the docs

File: scripts/init.py
Comment (jjgriff93) [reply scripts/init.py:1]: Why on earth are you creating a yaml config file from a tfvars file???

File: scripts/init.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        assert value != "__CHANGE_ME__", f"Value of {key} must be defined"
+
+        return f'{key:40s} = "{value}"'
+
+    with open(Path(REPO_ROOT_DIR, "config.tfvars"), "w") as config_file:
+        for line in open(Path(REPO_ROOT_DIR, "config.sample.tfvars"), "r"):
+            print(_processed(line), file=config_file)
+
+
+@log("Modifying .gitignore to allow checkin config and state")
```
Comment (jjgriff93) [inline scripts/init.py:66]: Can't resist a good script can you - this seems like overkill when we can just say "remove a line from gitignore if you wish to check in state" in the readme xD

File: scripts/modify_ip_exceptions.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+set -o errexit
+set -o pipefail
+set -o nounset
+# set -o xtrace
+
```
Comment (jjgriff93) [inline scripts/modify_ip_exceptions.sh:20]: Now this script I like ;)

File: .gitignore (hunk 2 above)
Comment (jjgriff93) [reply .gitignore:30]: Have seen your script further down - filled me with dread

File: deployment/build-agent/cloud-config.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjgriff93) [reply deployment/build-agent/cloud-config.yaml:54]: Then let's do that :D guess I meant does it add any extra issues by not doing it as root

File: scripts/az_login.sh [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+set -o errexit
+set -o pipefail
+set -o nounset
+
+SUB_NAME=$(az account show --query name -o tsv)
+TENANT_ID=$(az account show --query tenantId -o tsv)
+
+if [ -z "$SUB_NAME" ]; then
+  echo -e "\n\e[31m»»» ⚠️ You are not logged in to Azure!"
```
[the remark this answers is not in the document]
Comment (jjgriff93) [reply scripts/az_login.sh:24]: I shall get rid ;)

**Pull request merged**